zcash_client_backend/CHANGELOG.md

```diff
@@ -20,6 +20,11 @@ and this library adheres to Rust's notion of
 
 ### Changed
 - MSRV is now 1.70.0.
+- `zcash_client_backend::{fixed,standard,zip317}::SingleOutputChangeStrategy`
+  now implement a slightly different strategy for choosing whether there will
+  be any change. Transactions that have any shielded non-change flows will
+  always have a change output (possibly zero-valued). This can avoid leaking
+  information about note amounts in some cases.
```

Comment:
It should be considered whether this subsumes padding at the transaction builder level. I'd personally prefer it if the low-level transaction builders weren't responsible for padding; that would make for a much better separation of concerns and would eliminate the kind of leakage I think you're talking about here: the builder padding can actually reveal the absence of change if the recipient of a transaction is able to determine (via collusion with some other parties) that both outputs are not change.

Reply:
It might, but I'd rather consider that in another PR at this point.

Reply:
> the builder padding can actually reveal the absence of change if the recipient of a transaction is able to determine (via collusion with some other parties) that both outputs are not change.

This is a tautology. If the recipient is able to determine via collusion that both outputs of a 2-output transaction are not change, then they have learned that neither output is change, and would have learned this regardless of how the builder pads.

The builder padding provides arity hiding against a third-party blockchain observer, and against a non-colluding recipient. That is, the recipient is _required_ to collude in order to distinguish whether they are the sole recipient in a 2-output transaction, whereas without builder padding no collusion is required.

Reply:
The behaviour alteration made in this PR *only* affects the situation where there is exact change for a fully or partially shielded transaction. This previously could leak information about whether there is exact change, as long as the adversary has any information at all about the distribution of the number of outputs prior to adding the change output (which they do: the distribution is non-uniform).

The builder padding happens at a lower level. It can't prevent the information leak: the current policy could only even potentially do that when the original number of outputs (for Sapling) or actions (for Orchard) before adding the change output is fewer than 2.

(In the case of Orchard, if the number of inputs is at least one greater than the number of outputs then the presence of a change output will be hidden regardless of this behaviour alteration.)

Reply:
> The builder padding provides arity hiding against a third-party blockchain observer, and against a non-colluding recipient. That is, the recipient is _required_ to collude in order to distinguish whether they are the sole recipient in a 2-output transaction, whereas without builder padding no collusion is required.

My point here is that the padding would still occur; it would just occur in the higher-level transaction construction methods rather than in the low-level builder. In any case, it's not important for this PR.